torch/_higher_order_ops/cond.py

```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
         ignore_fresh_unbacked = mode.shape_env.ignore_fresh_unbacked_symbols()
 
     with mode, ignore_fresh_unbacked:
-        true_outs = true_fn(*operands)
-        flat_true_outs = pytree.tree_leaves(true_outs)
-        flat_false_outs = pytree.tree_leaves(false_fn(*operands))
-    if len(flat_true_outs) != len(flat_false_outs):
-        raise RuntimeError("Unmatched number of outputs from cond() branches.")
+        flat_true_outs, true_out_spec = pytree.tree_flatten(true_fn(*operands))
+        flat_false_outs, false_out_spec = pytree.tree_flatten(false_fn(*operands))
+        if true_out_spec != false_out_spec:
+            raise RuntimeError(
+                "Unmatched output spec from torch.cond branches: "
+                f"true branch tree_spec {true_out_spec} vs false branch tree_spec {false_out_spec}."
+            )
 
+    merged_outs = []
     for true_out, false_out in zip(flat_true_outs, flat_false_outs):
-        if true_out is None or false_out is None:
-            if true_out is None and false_out is None:
-                continue
-            raise torch._dynamo.exc.CondOpArgsMismatchError(
-                f"Expected both branches to return None:"
-                f"\n  {true_fn.__name__} returns {true_out}"
-                f"\n  {false_fn.__name__} returns {false_out}"
+        merged_outs.append(_merge_tensors(true_out, false_out, mode))
+    return pytree.tree_unflatten(merged_outs, true_out_spec)
+
+
+def _merge_tensors(a: torch.Tensor, b: torch.Tensor, mode: FakeTensorMode):
+    from torch.fx.experimental.symbolic_shapes import (
+        _nested_int_aware_sort,
+        SymIntEqByExpr,
+    )
+    from torch.utils._sympy.value_ranges import bound_sympy, ValueRanges
+
+    if a is None or b is None:
+        assert a is None and b is None, (a, b)
+        return None
+    torch._check(
+        a.dtype == b.dtype,
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same dtype but got {a.dtype} and {b.dtype}",
+    )
+    torch._check(
+        a.device == b.device,
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same device but got {a.device} and {b.device}",
+    )
+    torch._check(
+        a.dim() == b.dim(),
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same dim but got {a.dim()} and {b.dim()}",
+    )
```

Comment:
Are these the only things to be checked? If not it might be nice to have a helper function that easily checks a single property and then we can add more properties to it.

There's also storage_offset, maybe more properties (like is_conj/is_neg, dispatchkeys, etc)